tests/csapi/push_test.go

```diff
@@ -41,3 +44,93 @@ func TestPushRuleCacheHealth(t *testing.T) {
+		pushrules := syncResp.Get(`account_data.events.#(type=="m.push_rules").content.global`)
+		if !pushrules.Exists() {
+			t.Fatalf("no pushrules found in sync response: %s", syncResp.Raw)
+		}
+	})
+
+	// sytest: Adding a push rule wakes up an incremental /sync
+	t.Run("Adding a push rule wakes up an incremental /sync", func(t *testing.T) {
+		nextBatch = checkWokenUp(t, alice, client.SyncReq{Since: nextBatch, TimeoutMillis: "10000"}, func() {
+			body := client.WithJSONBody(t, map[string]interface{}{
+				"actions": []string{"notify"},
+			})
+
+			alice.MustDoFunc(t, "PUT", []string{"_matrix", "client", "v3", "pushrules", "global", "room", "!foo:example.com"}, body)
+		})
+	})
+
+	// sytest: Disabling a push rule wakes up an incremental /sync
+	t.Run("Disabling a push rule wakes up an incremental /sync", func(t *testing.T) {
+		nextBatch = checkWokenUp(t, alice, client.SyncReq{Since: nextBatch, TimeoutMillis: "10000"}, func() {
+			body := client.WithJSONBody(t, map[string]interface{}{
+				"enabled": false,
+			})
+
+			alice.MustDoFunc(t, "PUT", []string{"_matrix", "client", "v3", "pushrules", "global", "room", "!foo:example.com", "enabled"}, body)
+		})
+	})
+
+	// sytest: Enabling a push rule wakes up an incremental /sync
+	t.Run("Enabling a push rule wakes up an incremental /sync", func(t *testing.T) {
+		nextBatch = checkWokenUp(t, alice, client.SyncReq{Since: nextBatch, TimeoutMillis: "10000"}, func() {
+			body := client.WithJSONBody(t, map[string]interface{}{
+				"enabled": true,
+			})
+
+			alice.MustDoFunc(t, "PUT", []string{"_matrix", "client", "v3", "pushrules", "global", "room", "!foo:example.com", "enabled"}, body)
+		})
+	})
+
+	// sytest: Setting actions for a push rule wakes up an incremental /sync
+	t.Run("Setting actions for a push rule wakes up an incremental /sync", func(t *testing.T) {
+		checkWokenUp(t, alice, client.SyncReq{Since: nextBatch, TimeoutMillis: "10000"}, func() {
+			body := client.WithJSONBody(t, map[string]interface{}{
+				"actions": []string{"dont_notify"},
+			})
+
+			alice.MustDoFunc(t, "PUT", []string{"_matrix", "client", "v3", "pushrules", "global", "room", "!foo:example.com", "actions"}, body)
+		})
+	})
+}
+
+func checkWokenUp(t *testing.T, csapi *client.CSAPI, syncReq client.SyncReq, fn func()) (nextBatch string) {
+	t.Helper()
+	errChan := make(chan error, 1)
+	syncStarted := make(chan struct{})
+	go func() {
+		var syncResp gjson.Result
+		syncStarted <- struct{}{}
+		syncResp, nextBatch = csapi.MustSync(t, syncReq)
+		// get the first result where the type is m.push_rules
+		pushrules := syncResp.Get(`account_data.events.#(type=="m.push_rules").content.global`)
+		if !pushrules.Exists() {
+			errChan <- fmt.Errorf("no pushrules found in sync response: %s", syncResp.Raw)
+			return
+		}
+		errChan <- nil
+	}()
+	// Wait for the sync to actually start, to avoid responding immediately
+	<-syncStarted
+	fn()
+
+	if err := <-errChan; err != nil {
+		t.Fatal(err)
+	}
```

Comment:
Is there a reason we are using a goroutine instead of calling `fn()`, then doing the sync?

Reply:
Thanks, that makes sense. The wait doesn't actually guarantee that the `/sync` has started before `fn()` completes, which confused me. Could we make it a little clearer?
Maybe `// Try to wait for the sync to actually start, so that we test wakeups`?